### ISSUE TITLE
[BUG] DeepCompile in ZeRO-1 fails to do the forward pass

### ISSUE THREAD

**Issue by HollowMan6** [deepspeedai/DeepSpeed#7229]
**Describe the bug**
It looks like DeepCompile might have some compatibility issues with flash attention, when doing the forward pass.

Error logs:

```logs
MemoryProfiling error /pytorch/build/aten/src/ATen/RegisterCUDA.cpp:18503: SymIntArrayRef expected to contain only concrete integers

While executing %view : [num_users=2] = call_function[target=torch.ops.aten.view.default](args = (%arg1_1, [%arg0_1]), kwargs = {})
Original traceback:
  File "flash_attn/bert_padding.py", line 114, in torch_dynamo_resume_in_unpad_input_at_114
    indices = torch.nonzero(all_masks.flatten(), as_tuple=False).flatten()

```

**To Reproduce**
I'm using [OpenRLHF](https://github.com/OpenRLHF/OpenRLHF) with DeepCompile, but I think this should not be directly related to code in OpenRLHF and should be general enough to reproduce.

Basically, you may be able to reconstruct a reproduction script that, during training, run a forward pass when doing `.eval()` in ZeRO-1 with DeepCompile, for a huggingface model (`meta-llama/Llama-3.2-1B-Instruct`) with flash attention enabled (`transformers==4.51.1`, `flash_attn==2.7.4.post1`).

**Expected behavior**
No error thrown

**ds_report output**

```logs
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-devel package with yum
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
fused_adam ............. [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_lion ............... [NO] ....... [OKAY]
dc ..................... [NO] ....... [OKAY]
 [WARNING]  Please specify the CUTLASS repo directory as environment variable $CUTLASS_PATH
evoformer_attn ......... [NO] ....... [NO]
 [WARNING]  FP Quantizer is using an untested triton version (3.2.0), only 2.3.(0, 1) and 3.0.0 are known to be compatible with these kernels
fp_quantizer ........... [NO] ....... [NO]
fused_lamb ............. [NO] ....... [OKAY]
fused_lion ............. [NO] ....... [OKAY]
 [WARNING]  gds requires the dev libaio .so object and headers but these were not found.
 [WARNING]  gds: please install the libaio-devel package with yum
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
gds .................... [NO] ....... [NO]
transformer_inference .. [NO] ....... [OKAY]
inference_core_ops ..... [NO] ....... [OKAY]
cutlass_ops ............ [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
ragged_device_ops ...... [NO] ....... [OKAY]
ragged_ops ............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.6
 [WARNING]  using untested triton version (3.2.0), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch version .................... 2.6.0+cu124
deepspeed info ................... 0.16.6, unknown, unknown
torch cuda version ............... 12.4
torch hip version ................ None
nvcc version ..................... 12.4
deepspeed wheel compiled w. ...... torch 2.6, cuda 12.4
shared memory (/dev/shm) size .... 251.75 GB
```

**System info (please complete the following information):**
 - OS: RHEL 9.5
 - 1 machines with x4 A100s each
 - transformers==4.51.1
 - flash_attn==2.7.4.post1
 - Python version: 3.12.9
 - OpenRLHF version: 0.7.1.post1

**Comment by HollowMan6** [comment]
Error with flash attention disabled:

```logs
MemoryProfiling error /pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:2352: SymIntArrayRef expected to contain only concrete integers

While executing %expand_1 : [num_users=1] = call_function[target=torch.ops.aten.expand.default](args = (%unsqueeze_5, [%primals_1, 1, -1, -1]), kwargs = {})
Original traceback:
  File "openrlhf/models/model.py", line 266, in forward
    outputs = getattr(self, self.base_model_prefix)(
  File "transformers/utils/generic.py", line 965, in wrapper
    output = func(self, *args, **kwargs)
  File "transformers/models/llama/modeling_llama.py", line 541, in forward
    causal_mask = self._update_causal_mask(
  File "transformers/models/llama/modeling_llama.py", line 647, in _update_causal_mask
    causal_mask = self._prepare_4d_causal_attention_mask_with_cache_position(
  File "transformers/models/llama/modeling_llama.py", line 717, in _prepare_4d_causal_attention_mask_with_cache_position
    causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
```

**Comment by tohtana** [comment]
Hi @HollowMan6,

Thank you for reporting this!

I tried to match the versions and configurations with yours:
- PyTorch v2.6
- transformers 4.51.1
- flash_attn 2.7.4.post1
But I wasn't able to reproduce this issue on my end.

I used an example to run DeepCompile here:
https://github.com/deepspeedai/DeepSpeedExamples/benchmarks/deepcompile

And I ran the following command:

```
bash ./run.sh --host-ip localhost --backend deepspeed --model meta-llama/Llama-3.2-1B-Instruct --batch-size 1 --seq-length 512 --zero-stage 1 --gradient-accumulation-steps 1 --compile --deepcompile --attn-impl flash_attention_2 --eval
```

This enables DeepCompile with flash-attention and evaluation mode, and it runs without any errors.
It should be a difference in the configuration or input data. Could you share any details you can find?

**Comment by HollowMan6** [comment]
OpenRLHF didn't use `accelerate` but `ray`, and there is some customization code there to fit RL use cases.

Aside from that, OpenRLHF did some customization on the input data (with `packing_samples` when `flash_attn` is enabled https://github.com/deepspeedai/DeepSpeed/issues/7229#issue-3003383527 https://github.com/deepspeedai/DeepSpeed/issues/7228#issue-3003374006), https://github.com/OpenRLHF/OpenRLHF/blob/52498da5fd476ad12393e885590c439baa90672c/openrlhf/models/actor.py#L181-L185, before passing to the deepspeed engine.

When `flash_attn` is disabled (`packing_samples` is also disabled) https://github.com/deepspeedai/DeepSpeed/issues/7229#issuecomment-2814147263, according to the code comment here, the provided `attention` mask should be 2D. https://github.com/huggingface/transformers/blob/8ab296501a2e4c6ddada77118acd29c7c61c7795/src/transformers/models/llama/modeling_llama.py#L652

**Comment by tohtana** [comment]
@HollowMan6 Thank you for the additional information.

I double-checked that we also used a 2D mask and passed this line, but haven't still reproduced the error.
```
File "transformers/models/llama/modeling_llama.py", line 717, in _prepare_4d_causal_attention_mask_with_cache_position
    causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
```

I will try running OpenRLHF, but it might need some time. I would also appreciate it if you could try other settings in OpenRLHF, e.g. disabling `packing_samples`.

**Comment by HollowMan6** [comment]
Thanks again for investigation @tohtana ! The packing samples option has already been disabled when I disabled the flash attention (as the packing samples option requires flash attention). I'm not sure if it's because OpenRLHF wrapped on top of DeepSpeed engine while only DeepSpeed engine gets compiled that caused this error or not, but would be appreciated if you could share some insights here!

For your reference, the entry point for initialization of DeepSpeed is at https://github.com/OpenRLHF/OpenRLHF/blob/main/openrlhf/trainer/ray/ppo_actor.py#L338

The script I used for OpenRLHF when you have 8 GPUs, remove the last two lines to disable packing samples:

```bash
ray job submit \
   -- python3 -m openrlhf.cli.train_ppo_ray \
   --ref_num_nodes 1 \
   --ref_num_gpus_per_node 2 \
   --reward_num_nodes 1 \
   --reward_num_gpus_per_node 2 \
   --critic_num_nodes 1 \
   --critic_num_gpus_per_node 2 \
   --actor_num_nodes 1 \
   --actor_num_gpus_per_node 2 \
   --vllm_num_engines 4 \
   --vllm_tensor_parallel_size 1 \
   --colocate_critic_reward \
   --colocate_actor_ref \
   --pretrain meta-llama/Llama-3.2-1B-Instruct \
   --reward_pretrain meta-llama/Llama-3.2-1B-Instruct \
   --micro_train_batch_size 2 \
   --train_batch_size 16 \
   --micro_rollout_batch_size 4 \
   --rollout_batch_size 64 \
   --max_samples 1024 \
   --n_samples_per_prompt 1 \
   --bf16 \
   --max_epochs 1 \
   --prompt_max_len 1024 \
   --generate_max_len 1024 \
   --zero_stage 1 \
   --actor_learning_rate 5e-7 \
   --critic_learning_rate 9e-6 \
   --init_kl_coef 0.01 \
   --prompt_data OpenRLHF/prompt-collection-v0.1 \
   --input_key context_messages \
   --apply_chat_template \
   --normalize_reward \
   --overlap_comm \
   --deepcompile \
   --packing_samples \
   --flash_attn
```

**Comment by tohtana** [comment]
Hi @HollowMan6,
I did some fixes in DeepCompile, and it worked with ZeRO1 and OpenRLHF. Currently we have an issue with CI, but you can try the fixes in #7243.
We also need padding in OpenRLHF side and SDPF instead of FlashAttention. Please see this [branch](https://github.com/tohtana/OpenRLHF/tree/tohtana/dc_pad_seq) for the necessary changes. Using the command you shared, forward/backward of the actor and the critic was faster than the settings without DeepCompile.

We still have issues with ZeRO3 and FlashAttention. I will post here (or #7228) when I have update.

**Comment by HollowMan6** [comment]
@hijkzzz Would you double-check if there's any potential problem here for the fixes to get OpenRLHF supported for deepcompile in ZeRO-1 https://github.com/tohtana/OpenRLHF/tree/tohtana/dc_pad_seq, and also get this properly merged into OpenRLHF?

**Comment by tohtana** [comment]
@HollowMan6 @hijkzzz The padding used in my branch adds a small overhead to share the maximum length across ranks. It also forcibly uses SDPA and disables packing for the critic and reward models. Please address these properly when you merge them.

**Comment by hijkzzz** [comment]
Has this bug been fixed?

**Comment by hijkzzz** [comment]
@tohtana This bug was not solved when using flash-attn-2.7.4.post1 and deepspeed 0.17.1.

**Comment by tohtana** [comment]
@HollowMan6 @hijkzzz Thank you for checking! Yes, we still have a few issues. Let me summarize the status:

- We still haven't addressed the issue with flash attention.
- #7243 fixed an issue with dynamic sequence lengths. I'm now able to run OpenRLHF using SDPF (with [my custom branch for seq padding](https://github.com/tohtana/OpenRLHF/tree/tohtana/dc_pad_seq). This shows a certain speedup of forward with Llama-3-8B on my environment, but also shows slowdown for backward. I'm investigating it now.

I will be able to investigate more on flash attention after I resolve the slowdown of backward. I will keep you posted here.

**Comment by tohtana** [comment]
@HollowMan6 @hijkzzz 

Let me share some update:
- PyTorch's compiler doesn't work with flash attention when the mask is set via sequence packing. It is not related to DeepSpeed or DeepCompile. You can use SDPA instead.
  - Please refer to this [verification](https://github.com/tohtana/actor_debug_compile) for details
  - Let us leave this issue until PyTorch compiler or Flash Attention is updated
- #7366 improves ZeRO1's performance
  - I enabled DeepCompile only for Actor 
  - See the [benchmark](https://github.com/tohtana/DeepCompile_docs/wiki/Debug-with-OpenRLHF-(%237243)#benchmark)
  - This PR will be merged after the review is done

You can see the summary at [Wiki page](https://github.com/tohtana/DeepCompile_docs/wiki/Debug-with-OpenRLHF-(%237243)).
